// @grafana/faro-web-sdk/dist/cjs/instrumentations/errors/stackFrames/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reactMinifiedRegex = exports.safariWebExtensionString = exports.safariExtensionString = exports.firefoxEvalString = exports.firefoxEvalRegex = exports.firefoxLineRegex = exports.webkitAddressAtStringLength = exports.webkitAddressAtString = exports.webkitEvalString = exports.webkitEvalRegex = exports.webkitLineRegex = exports.atString = exports.unknownSymbolString = exports.evalString = exports.newLineString = void 0;
exports.newLineString = '\n';
exports.evalString = 'eval';
exports.unknownSymbolString = '?';
exports.atString = '@';
exports.webkitLineRegex = /^\s*at (?:(.*?) ?\((?:address at )?)?((?:file|https?|blob|chrome-extension|address|native|eval|webpack|<anonymous>|[-a-z]+:|.*bundle|\/)?.*?)(?::(\d+))?(?::(\d+))?\)?\s*$/i;
exports.webkitEvalRegex = /\((\S*)(?::(\d+))(?::(\d+))\)/;
exports.webkitEvalString = 'eval';
exports.webkitAddressAtString = 'address at ';
exports.webkitAddressAtStringLength = exports.webkitAddressAtString.length;
exports.firefoxLineRegex = /^\s*(.*?)(?:\((.*?)\))?(?:^|@)?((?:file|https?|blob|chrome|webpack|resource|moz-extension|safari-extension|safari-web-extension|capacitor)?:\/.*?|\[native code]|[^@]*(?:bundle|\d+\.js)|\/[\w\-. /=]+)(?::(\d+))?(?::(\d+))?\s*$/i;
exports.firefoxEvalRegex = /(\S+) line (\d+)(?: > eval line \d+)* > eval/i;
exports.firefoxEvalString = ' > eval';
exports.safariExtensionString = 'safari-extension';
exports.safariWebExtensionString = 'safari-web-extension';
exports.reactMinifiedRegex = /Minified React error #\d+;/i;
