// @grafana/schema/dist/esm/raw/composable/testdata/dataquery/x/TestDataDataQuery_types.gen.js
const pluginVersion = "%VERSION%";
var TestDataQueryType = /* @__PURE__ */ ((TestDataQueryType2) => {
  TestDataQueryType2["Annotations"] = "annotations";
  TestDataQueryType2["Arrow"] = "arrow";
  TestDataQueryType2["CSVContent"] = "csv_content";
  TestDataQueryType2["CSVFile"] = "csv_file";
  TestDataQueryType2["CSVMetricValues"] = "csv_metric_values";
  TestDataQueryType2["DataPointsOutsideRange"] = "datapoints_outside_range";
  TestDataQueryType2["ExponentialHeatmapBucketData"] = "exponential_heatmap_bucket_data";
  TestDataQueryType2["FlameGraph"] = "flame_graph";
  TestDataQueryType2["GrafanaAPI"] = "grafana_api";
  TestDataQueryType2["LinearHeatmapBucketData"] = "linear_heatmap_bucket_data";
  TestDataQueryType2["Live"] = "live";
  TestDataQueryType2["Logs"] = "logs";
  TestDataQueryType2["ManualEntry"] = "manual_entry";
  TestDataQueryType2["NoDataPoints"] = "no_data_points";
  TestDataQueryType2["NodeGraph"] = "node_graph";
  TestDataQueryType2["PredictableCSVWave"] = "predictable_csv_wave";
  TestDataQueryType2["PredictablePulse"] = "predictable_pulse";
  TestDataQueryType2["RandomWalk"] = "random_walk";
  TestDataQueryType2["RandomWalkTable"] = "random_walk_table";
  TestDataQueryType2["RandomWalkWithError"] = "random_walk_with_error";
  TestDataQueryType2["RawFrame"] = "raw_frame";
  TestDataQueryType2["ServerError500"] = "server_error_500";
  TestDataQueryType2["Simulation"] = "simulation";
  TestDataQueryType2["SlowQuery"] = "slow_query";
  TestDataQueryType2["StreamingClient"] = "streaming_client";
  TestDataQueryType2["TableStatic"] = "table_static";
  TestDataQueryType2["Trace"] = "trace";
  TestDataQueryType2["USA"] = "usa";
  TestDataQueryType2["VariablesQuery"] = "variables-query";
  return TestDataQueryType2;
})(TestDataQueryType || {});
const defaultUSAQuery = {
  fields: [],
  states: []
};
const defaultTestDataDataQuery = {
  csvWave: [],
  points: [],
  scenarioId: "random_walk" /* RandomWalk */
};

export { TestDataQueryType, defaultTestDataDataQuery, defaultUSAQuery, pluginVersion };
