// @grafana/faro-core/dist/spec/core/src/internalLogger/const.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultInternalLoggerLevel = exports.defaultInternalLogger = exports.defaultInternalLoggerPrefix = exports.InternalLoggerLevel = void 0;
var utils_1 = require("../utils");
var InternalLoggerLevel;
(function (InternalLoggerLevel) {
    InternalLoggerLevel[InternalLoggerLevel["OFF"] = 0] = "OFF";
    InternalLoggerLevel[InternalLoggerLevel["ERROR"] = 1] = "ERROR";
    InternalLoggerLevel[InternalLoggerLevel["WARN"] = 2] = "WARN";
    InternalLoggerLevel[InternalLoggerLevel["INFO"] = 3] = "INFO";
    InternalLoggerLevel[InternalLoggerLevel["VERBOSE"] = 4] = "VERBOSE";
})(InternalLoggerLevel || (exports.InternalLoggerLevel = InternalLoggerLevel = {}));
exports.defaultInternalLoggerPrefix = 'Faro';
exports.defaultInternalLogger = {
    debug: utils_1.noop,
    error: utils_1.noop,
    info: utils_1.noop,
    prefix: exports.defaultInternalLoggerPrefix,
    warn: utils_1.noop,
};
exports.defaultInternalLoggerLevel = InternalLoggerLevel.ERROR;
