// @grafana/faro-core/dist/cjs/transports/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransportBody = exports.registerInitialTransports = exports.transportItemTypeToBodyKey = exports.TransportItemType = exports.BaseTransport = exports.initializeTransports = void 0;
var initialize_1 = require("./initialize");
Object.defineProperty(exports, "initializeTransports", { enumerable: true, get: function () { return initialize_1.initializeTransports; } });
var base_1 = require("./base");
Object.defineProperty(exports, "BaseTransport", { enumerable: true, get: function () { return base_1.BaseTransport; } });
var const_1 = require("./const");
Object.defineProperty(exports, "TransportItemType", { enumerable: true, get: function () { return const_1.TransportItemType; } });
Object.defineProperty(exports, "transportItemTypeToBodyKey", { enumerable: true, get: function () { return const_1.transportItemTypeToBodyKey; } });
var registerInitial_1 = require("./registerInitial");
Object.defineProperty(exports, "registerInitialTransports", { enumerable: true, get: function () { return registerInitial_1.registerInitialTransports; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "getTransportBody", { enumerable: true, get: function () { return utils_1.getTransportBody; } });
