// @grafana/ui/dist/ReactMonacoEditor-a6bb6d65.js
'use strict';

var MonacoEditor = require('@monaco-editor/react');
var React = require('react');
require('@grafana/data');
var index = require('./index.js');
require('micro-memoize');
require('@emotion/react');
var tinycolor = require('tinycolor2');
require('@emotion/css');
require('classnames');
require('react-inlinesvg');
require('hoist-non-react-statics');
require('@floating-ui/react');
require('@grafana/e2e-selectors');
require('react-dom');
require('lodash');
require('slate');
require('ansicolor');
require('react-select');
require('react-select/async');
require('react-select/creatable');
require('react-custom-scrollbars-2');
require('react-transition-group');
require('react-popper');
require('react-transition-group/Transition');
require('i18next');
require('react-i18next');
require('rc-cascader');
require('react-use/lib/useMeasure');
require('react-select/async-creatable');
require('react-window');
require('@react-aria/focus');
require('react-colorful');
require('react-use');
require('date-fns');
require('@react-aria/dialog');
require('@react-aria/overlays');
require('react-calendar');
require('uuid');
require('rc-time-picker');
require('react-table');
require('react-loading-skeleton');
require('slate-plain-serializer');
require('slate-react');
require('is-hotkey');
require('prismjs');
require('immutable');
require('calculate-size');
require('react-highlight-words');
require('react-router-dom');
require('rxjs');
require('rxjs/operators');
require('@grafana/schema');
require('uplot');
require('jquery');
require('react-dom/server');
require('react-use/lib/usePrevious');
require('react-use/lib/useClickAway');
require('@grafana/faro-web-sdk');
require('rc-drawer');
require('rc-slider');
require('rc-tooltip');
require('react-hook-form');
require('react-dropzone');
require('ol/format/WKT');
require('ol/geom');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var MonacoEditor__default = /*#__PURE__*/_interopDefaultLegacy(MonacoEditor);
var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var tinycolor__default = /*#__PURE__*/_interopDefaultLegacy(tinycolor);

function getColors(theme) {
  if (theme === void 0) {
    return {};
  } else {
    const colors = {
      "editor.background": theme.components.input.background,
      "minimap.background": theme.colors.background.secondary
    };
    Object.keys(colors).forEach((resultKey) => {
      colors[resultKey] = normalizeColorForMonaco(colors[resultKey]);
    });
    return colors;
  }
}
function normalizeColorForMonaco(color) {
  return tinycolor__default["default"](color).toHexString();
}
function defineThemes(monaco, theme) {
  const colors = getColors(theme);
  monaco.editor.defineTheme("grafana-dark", {
    base: "vs-dark",
    inherit: true,
    colors,
    // fallback syntax highlighting for languages that microsoft doesn't handle (ex cloudwatch's metric math)
    rules: [
      { token: "predefined", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("purple")) },
      { token: "operator", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("orange")) },
      { token: "tag", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("green")) }
    ]
  });
  monaco.editor.defineTheme("grafana-light", {
    base: "vs",
    inherit: true,
    colors,
    // fallback syntax highlighting for languages that microsoft doesn't handle (ex cloudwatch's metric math)
    rules: [
      { token: "predefined", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("purple")) },
      { token: "operator", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("orange")) },
      { token: "tag", foreground: normalizeColorForMonaco(theme == null ? void 0 : theme.visualization.getColorByName("green")) }
    ]
  });
}

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
var _a;
MonacoEditor.loader.config({
  paths: {
    vs: ((_a = window.__grafana_public_path__) != null ? _a : "public/") + "lib/monaco/min/vs"
  }
});
const ReactMonacoEditor = (props) => {
  const { beforeMount } = props;
  const theme = index.useTheme2();
  const onMonacoBeforeMount = React.useCallback(
    (monaco) => {
      defineThemes(monaco, theme);
      beforeMount == null ? void 0 : beforeMount(monaco);
    },
    [beforeMount, theme]
  );
  return /* @__PURE__ */ React__default["default"].createElement(
    MonacoEditor__default["default"],
    __spreadProps(__spreadValues({}, props), {
      theme: theme.isDark ? "grafana-dark" : "grafana-light",
      beforeMount: onMonacoBeforeMount
    })
  );
};

exports.ReactMonacoEditor = ReactMonacoEditor;
