// @grafana/faro-core/dist/spec/core/src/semantic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EVENT_ROUTE_CHANGE = exports.EVENT_OVERRIDES_SERVICE_NAME = exports.EVENT_SESSION_EXTEND = exports.EVENT_SESSION_RESUME = exports.EVENT_SESSION_START = exports.EVENT_VIEW_CHANGED = exports.EVENT_NAVIGATION = exports.EVENT_CLICK = exports.Conventions = void 0;
/**
 * @deprecated The conventions object will be removed in a future version
 */
exports.Conventions = {
    /**
     * @deprecated The event names object will be removed in a future version
     */
    EventNames: {
        CLICK: 'click',
        NAVIGATION: 'navigation',
        SESSION_START: 'session_start',
        VIEW_CHANGED: 'view_changed',
    },
};
exports.EVENT_CLICK = 'click';
exports.EVENT_NAVIGATION = 'navigation';
exports.EVENT_VIEW_CHANGED = 'view_changed';
exports.EVENT_SESSION_START = 'session_start';
exports.EVENT_SESSION_RESUME = 'session_resume';
exports.EVENT_SESSION_EXTEND = 'session_extend';
exports.EVENT_OVERRIDES_SERVICE_NAME = 'service_name_override';
exports.EVENT_ROUTE_CHANGE = 'route_change';
